src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -0,0 +1,1093 @@
+            //               curr_min.minRank,
+            //               query_inds[idx],            // index
+            //               curr_min.minElem));
+
+            query_inds[idx] = curr_min.minElem;
+            query_ranks[idx] = curr_min.minRank;
+            query_pos[idx] = m_points[curr_min.minElem];
+
+            //DEBUG
+            query_min_dist[idx] = sqrt(curr_min.minSqDist);
+          }
+          // else
+          // {
+          //   SLIC_INFO(
+          //     fmt::format("[On rank {}] Didn't update dist for particle {} ::"
+          //                 "{{ cp: {}; "
+          //                 "dist: {}; "
+          //                 "rank: {}; "
+          //                 "elem: {}}}",
+          //                 rank,
+          //                 idx,
+          //                 query_pos[idx],
+          //                 query_min_dist[idx],
+          //                 query_ranks[idx],
+          //                 query_inds[idx]));
+          // }
+        }););
+
+    axom::copy(cpIndexes.data(),
+               query_inds.data(),
+               cpIndexes.size() * sizeof(axom::IndexType));
+    axom::copy(cpRanks.data(),
+               query_ranks.data(),
+               cpRanks.size() * sizeof(axom::IndexType));
+    axom::copy(closestPts.data(),
+               query_pos.data(),
+               closestPts.size() * sizeof(PointType));
+
+    // DEBUG
+    axom::copy(minDist.data(),
+               query_min_dist.data(),
+               minDist.size() * sizeof(double));
+  }
+
+private:
+  RuntimePolicy m_runtimePolicy;
+  bool m_isVerbose {false};
+  int m_allocatorID;
+  int m_rank;
+  int m_nranks;
+
+  PointArray m_points;
+  BoxArray m_boxes;
+
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+  SeqBVHTree* m_bvh_seq {nullptr};
+
+  #ifdef AXOM_USE_OPENMP
+  OmpBVHTree* m_bvh_omp {nullptr};
+  #endif
+
+  #ifdef AXOM_USE_CUDA
+  CudaBVHTree* m_bvh_cuda {nullptr};
+  #endif
+#endif
+};
+
+}  // namespace internal
+
+/**
+ * \brief Encapsulated the Distributed closest point query for a collection of query points 
+ * over an "object mesh"
+ *
+ * The object mesh and the query mesh are provided as conduit nodes using the mesh blueprint schema.
+ * Each of these are distributed over the problem's mpi ranks. This class orchestrates passing
+ * the query points to all ranks whose object meshes might contain a closest point.
+ *
+ * \note The class currently supports object meshes that are comprised of a collection of points.
+ * In the future, we'd like to consider more general object meshes, e.g. triangle meshes.
+ *
+ * To use this class, first set some parameters, such as the runtime execution policy,
+ * then pass in the object mesh and build a spatial index over this mesh.
+ * Finally, compute the closest points in the object mesh to each point in a query mesh 
+ * using the \a computeClosestPoint() function.
+ */
```

Comment:
Please note mesh assumptions and preconditions, including that the object and point meshes should be decomposed over the same number of domains, and that the work is decomposed along the object mesh decomposition.  No redecomposition is done.  Thank you for the chat where we discussed this.

Reply:
Thanks. I also added a note that the class will need to support the case where a rank has an empty object mesh (i.e. no object points), but that this is not currently supported/tested in this initial version.